pkg/sfsturbo/controllerserver.go

```
@@ -18,179 +18,425 @@ package sfsturbo
-    if err != nil {
-		klog.V(3).Infof("Failed to create SFS Turbo v1 client: %v", err)
-		return nil, status.Error(codes.Internal, err.Error())
-    }
-
-	requestedSize := req.GetCapacityRange().GetRequiredBytes()
-    if requestedSize == 0 {
-        // At least 500GiB
-        requestedSize = 500 * tInGiB
-    }
-
-    sizeInGiB := bytesToGiB(requestedSize)
-
-	// Creating a share
-	createOpts := shares.CreateOpts{}
-	createOpts.VpcID = cs.Driver.cloud.Vpc.Id
-	createOpts.SecurityGroupID = cs.Driver.cloud.Vpc.SecurityGroupId
-	createOpts.SubnetID = cs.Driver.cloud.Vpc.SubnetId
-	createOpts.AvailabilityZone = cs.Driver.cloud.Global.AvailabilityZone
-	// build name
-	createOpts.Name = req.GetName()
-	// build share proto
-	createOpts.ShareProto = cs.Driver.shareProto
-	createOpts.Size = sizeInGiB
-	// build type
-	createOpts.ShareType = cs.Driver.cloud.Ext.ShareProto
-	if createOpts.ShareType == "" {
-		createOpts.ShareType = "STANDARD"
-		klog.V(2).Infof("Creating params STANDARD default")
-	}
-	klog.V(2).Infof("Creating params: %v", createOpts)
-
-    share, err := createShare(client, &createOpts)
+const (
+	defaultShareProto = "NFS"
+	defaultShareType  = "STANDARD"
+	minSizeInGiB      = 500
+	maxSizeInGiB      = 32768
+	expandStepInGiB   = 100
+)
+
+// resourcemode from SC
+const (
+	Availability = "availability"
+	ShareType    = "shareType"
+)
+
+func (cs *controllerServer) CreateVolume(_ context.Context, req *csi.CreateVolumeRequest) (
+	*csi.CreateVolumeResponse, error) {
+	log.Infof("CreateVolume called with request %v", protosanitizer.StripSecrets(*req))
+	cloud := cs.Driver.cloud
+
+	name := req.GetName()
+	capacityRange := req.GetCapacityRange()
+	if err := createVolumeValidation(name, capacityRange); err != nil {
+		return nil, err
+	}
+	sizeInGiB := int(utils.RoundUpSize(capacityRange.GetRequiredBytes(), common.GbByteSize))
+	// 500 ~ 32768
+	if sizeInGiB < minSizeInGiB {
+		sizeInGiB = minSizeInGiB
+	} else if sizeInGiB > maxSizeInGiB {
```

exceeding the maximum capacity should return an error, do not create a small one

---

done